SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/SpatialPingComponent.h"
+
+#include "GameFramework/PlayerController.h"
+#include "Net/UnrealNetwork.h"
+#include "TimerManager.h"
+#include "Engine/World.h"
+
+USpatialPingComponent::USpatialPingComponent(const FObjectInitializer& ObjectInitializer) : Super(ObjectInitializer)
+{
+	PrimaryComponentTick.bCanEverTick = false;
+	PrimaryComponentTick.bStartWithTickEnabled = false;
+
+	bReplicates = true;
+
+	bStartWithPingEnabled = true;
+}
+
+void USpatialPingComponent::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(USpatialPingComponent, ReplicatedPingID);
+}
+
+void USpatialPingComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	//Attempt to cast the owner of this component to the PlayerController class
+	//Component will do nothing if cast fails
+	OwningController = Cast<APlayerController>(GetOwner());
+	if (bStartWithPingEnabled)
+		SetPingEnabled(true);
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	Super::EndPlay(EndPlayReason);
+	//Clear up the timer before the component is disabled or destroyed
+	SetPingEnabled(false);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	if (OwningController) {
+		//Only execute on owning local client
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled) {
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled) {
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RTPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	UWorld* World = GetWorld();
+	if (World)
+	{
+		//Set looping timer to 'tick' this component and send ping RPC with configurable frequency
+		World->GetTimerManager().SetTimer(PingTimerHandle, this, &USpatialPingComponent::TickPingComponent, PingFrequency, true);
+		bIsPingEnabled = true;
+	}
+}
+
+void USpatialPingComponent::DisablePing()
+{
+	UWorld* World = GetWorld();
+	if (World)
+	{
+		//Clear the timer
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+		bIsPingEnabled = false;
+		RTPing = 0.f;
+	}
+}
+
+void USpatialPingComponent::TickPingComponent()
+{
+	SendNewPing();
+	//Pass latest measured ping to owning controller to be processed by PlayerState
+	OwningController->UpdatePing(RTPing);
+}
+
+void USpatialPingComponent::SendNewPing()
+{
+	//Send a new ping using the current local time since start as both ID and timestamp
+	SendServerWorkerPingID(GetWorld()->GetRealTimeSeconds());
```

Comment:
We should probably use `FDateTime::Now` instead. The problem with `UWorld::GetRealTimeSeconds` is it's updated every tick, and so doesn't necessarily represent the latency. With native networking, I ran into a situation where the client received the replicated property from the server on the same tick as it sent the RPC, so the recorded latency would be 0, which is hardly helpful.

Reply:
Yeah this makes sense, thanks. I'll replace that.

Reply:
I found that FPlatformTime::Seconds offered better resolution on the final output so I opted to use that instead. I believe that still uses the system clock so still resolves this problem.

Reply:
If in doubt there is also std::chrono if you need real time rather than game time.